staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/validation/validation.go

```diff
@@ -88,6 +88,38 @@ func ValidateCustomResourceDefinitionSpec(spec *apiextensions.CustomResourceDefi
 		allErrs = append(allErrs, field.NotSupported(fldPath.Child("scope"), spec.Scope, []string{string(apiextensions.ClusterScoped), string(apiextensions.NamespaceScoped)}))
 	}
 
+	if len(spec.Versions) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("versions"), ""))
+	} else {
+		storageFlagCount := 0
+		servedFlagCount := 0
+		servedVersions := map[string]bool{}
+		for _, version := range spec.Versions {
+			if version.Storage {
+				storageFlagCount++
+			}
+			if version.Served {
+				servedFlagCount++
+			}
+			if version.Served {
+				if servedVersions[version.Name] {
+					allErrs = append(allErrs, field.Invalid(fldPath.Child("versions"), spec.Versions, "duplicate version name "+version.Name))
+				} else {
+					servedVersions[version.Name] = true
+				}
+			}
+		}
+		if storageFlagCount != 1 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("versions"), spec.Versions, "one and only one version should be marked as storage version"))
+		}
+		if servedFlagCount == 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("versions"), spec.Versions, "at least one version should be served"))
+		}
+		if !servedVersions[spec.Version] {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("version"), spec.Version, "spec.version should be one of the items in spec.Versions"))
```

Comment:
The godoc says it must be the first, not "one of"

Reply:
Yes, but should it be first one? I know I documented it that way but I am just questioning that :)

Reply:
What else is the meaning of the version field otherwise? I would prefer to fix the meaning.

Reply:
I am not sure what you are suggesting here.